xbmc/games/addons/playback/GameLoop.cpp

```diff
@@ -58,79 +59,62 @@ void CGameLoop::Stop()
 
 void CGameLoop::SetSpeed(double speedFactor)
 {
-  {
-    CSingleLock lock(m_mutex);
-    m_speedFactor = speedFactor;
-    m_bPauseAsync = false;
-  }
+  m_speedFactor = speedFactor;
 
   m_sleepEvent.Set();
 }
 
 void CGameLoop::PauseAsync()
 {
-  {
-    CSingleLock lock(m_mutex);
-    m_bPauseAsync = true;
-  }
-
-  m_sleepEvent.Set();
+  SetSpeed(0.0);
 }
 
 void CGameLoop::Process(void)
 {
-  double nextFrameMs = NowMs();
-
-  CSingleLock lock(m_mutex);
-
   while (!m_bStop)
   {
-    double speedFactor = m_speedFactor;
-
-    // Pause game if the pause async flag was set
-    if (m_bPauseAsync)
+    if (m_speedFactor == 0.0)
     {
-      m_bPauseAsync = false;
-      m_speedFactor = 0.0;
+      m_lastFrameMs = 0.0;
+      m_sleepEvent.WaitMSec(5000);
     }
-
+    else
     {
-      CSingleExit exit(m_mutex);
-      if (speedFactor > 0.0)
+      if (m_speedFactor > 0.0)
         m_callback->FrameEvent();
-      else if (speedFactor < 0.0)
+      else if (m_speedFactor < 0.0)
         m_callback->RewindEvent();
-    }
 
-    // Record frame time
-    m_lastFrameMs = nextFrameMs;
+      if (m_lastFrameMs > 0.0)
+      {
+        m_lastFrameMs += FrameTimeMs();
+        m_adjustTime = m_lastFrameMs - NowMs();
+      }
+      else
+      {
+        m_lastFrameMs = NowMs();
+        m_adjustTime = 0.0;
+      }
+
+      CLog::Log(LOGDEBUG, "[RetroPlayer][GameLoop] adjust time = %0.4f", m_adjustTime);
```

Comment:
We should avoid logging for every frame before this hits master

Reply:
Agreed. Intended to remove this.